crypto/dh/dh_ameth.c

```diff
@@ -590,10 +587,12 @@ static void *dh_pkey_export_to(const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
     params = ossl_param_bld_to_param(&tmpl);
 
     /* We export, the provider imports */
-    provkey = evp_keymgmt_importkey(keymgmt, params);
+    provdata = want_domainparams
+        ? evp_keymgmt_importdomparams(keymgmt, params)
+        : evp_keymgmt_importkey(keymgmt, params);
```

Comment:
*Side note*: Having enforced a separation between provider side domain parameter objects and key objects, I can't help wondering whether there should actually be a provider side distinction between public keys and private keys....anyway  not for this PR!

Reply:
I did think about that distinction a while ago, but considering the private key *always* contains the public components as well, it seems futile and counter-productive.

Reply:
Basically, one can say that we always deal with key *pairs* but allow for the private parts to be absent.

Reply:
> the private key always contains the public components 

One could say the same thing about the public components and the domain parameters, i.e. "the public components always contain the domain parameters"